pingcap/tidb
# *: add tiflash replica sync progress

Signed-off-by: crazycs <crazycs520@gmail.com>



### What problem does this PR solve? 

Add tiflash replica sync `Progress` in `tiflash_replica` table.

```sql
information_schema> select table_name,available,progress from `TIFLASH_REPLICA`;
+------------+-----------+----------+
| table_name | available | progress |
+------------+-----------+----------+
| t          | 0         | 0.67     |
| tt         | 0         | 0.83     |
+------------+-----------+----------+
```

### What is changed and how it works?

Store the sync progress info in PD(etcd) when TiFlash report the progress.
When sync progress finished, delete the progress info in PD.


### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

Step 1: mock tiflash report sync info.
```shell
▶ curl -X POST -d '{"id":140,"region_count":5,"flash_region_count":6}' http://127.0.0.1:10080/tiflash/replica

▶ curl -X POST -d '{"id":129,"region_count":2,"flash_region_count":3}' http://127.0.0.1:10080/tiflash/replica
```

Step 2: check the progress info in the PD(etcd)
```shell
▶ etcdctl --endpoints=127.0.0.1:2379 get "/tiflash" --prefix
/tiflash/table/sync/129
0.67
/tiflash/table/sync/140
0.83
```

step 3: query progress info from `tiflash_replica`
```sql
information_schema> select table_name,table_id,available,progress from `TIFLASH_REPLICA`;
+------------+----------+-----------+----------+
| table_name | table_id | available | progress |
+------------+----------+-----------+----------+
| t          | 129      | 0         | 0.67     |
| tt         | 140      | 0         | 0.83     |
+------------+----------+-----------+----------+
```

Step 4: mock tiflash report sync info. mock for table_id=140 already sync finished.
```shell
▶ curl -X POST -d '{"id":140,"region_count":6,"flash_region_count":6}' http://127.0.0.1:10080/tiflash/replica
```

Step 5: check the progress info in the PD(etcd)
```shell
# the progress info of table_id=140 was been deleted. 
▶ etcdctl --endpoints=127.0.0.1:2379 get "/tiflash" --prefix
/tiflash/table/sync/129
0.67
```

step 6: query progress info from `tiflash_replica`
```sql
information_schema> select table_name,table_id,available,progress from `TIFLASH_REPLICA`;
+------------+----------+-----------+----------+
| table_name | table_id | available | progress |
+------------+----------+-----------+----------+
| t          | 129      | 0         | 0.67     |
| tt         | 140      | 1         | 1.0      |
+------------+----------+-----------+----------+
```


Code changes

 - Has exported function/method change

Side effects

Related changes

 - Need to cherry-pick to the release branch

Release note

 - Support query TiFlash replica synchronization progress in the `TIFLAH_REPLICA` table.
